src/ring_theory/dedekind_domain/integer_unit.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2022 David Kurniadi Angdinata. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Kurniadi Angdinata
+-/
+
+import ring_theory.dedekind_domain.adic_valuation
+
+/-!
+# `S`-integers and `S`-units of fraction fields of Dedekind domains
+
+This file defines the ring of `S`-integers and the group of `S`-units of the field of fractions of a
+Dedekind domain, to be specialised to the case of a number field or a function field separately.
+
+TODO: notation for `S`-integers and `S`-units, proof that `S`-integers is the intersection of
+valuation rings, proof that `S`-units is the kernel of a map to a product, proof that `∅`-integers
+is the usual ring of integers, finite generation of `S`-units and Dirichlet's `S`-unit theorem.
+
+## Main definitions
+
+* `is_dedekind_domain.S_integer_ring`: the ring of `S`-integers.
+* `is_dedekind_domain.S_unit_group`: the group of `S`-units.
+
+## Main statements
+
+* `is_dedekind_domain.S_integer_unit`: the units of the `S`-integers are the `S`-units.
+
+## References
+
+* https://encyclopediaofmath.org/wiki/S-integer
+* https://en.wikipedia.org/wiki/S-unit
+
+## Tags
+
+S-integer, S-unit
+-/
+
+open_locale non_zero_divisors
+
+namespace is_dedekind_domain
+
+noncomputable theory
+
+universes u v
+
+variables {R : Type u} [comm_ring R] [is_domain R] [is_dedekind_domain R] (K : Type v) [field K]
+  [algebra R K] [is_fraction_ring R K] (S : set $ height_one_spectrum R)
+
+/-- The subring of `S`-integers of `K`. -/
+def S_integer_ring : subring K :=
+{ carrier   := {x : K | ∀ v ∉ S, (v : height_one_spectrum R).valuation x ≤ 1},
+  mul_mem'  := λ _ _ hx hy v hv, by simp only [map_mul, mul_le_one₀ (hx v hv) (hy v hv)],
+  one_mem'  := λ _ _, by simp only [map_one, le_refl],
+  add_mem'  := λ _ _ hx hy v hv, v.valuation.map_add_le (hx v hv) (hy v hv),
+  zero_mem' := λ _ _, by simp only [map_zero, zero_le_one],
+  neg_mem'  := λ _ hx v hv, by simp only [valuation.map_neg, hx v hv] }
+
```

Comment:
This seems like it would be worth having:
```suggestion

/-- The `S`-integers of `R` are an algebra over `R`. -/
instance : algebra R (S_integer_ring K S) := -- your proof here

```